### PR TITLE
Gossip tuneables

### PR DESCRIPTION
**Pull request by mkeeler**
Fixes: #4058 

This PR exposes a number of the serf/memberlist/gossip tuneables. Previously all of these were being forced to default values. For users with extremely large clusters it would be nice to tune these to provide for quicker convergence and gossip message propagation (at the expense of increased network bandwidth).

The only changes here are to expose the configuration entries to the user. They end up just modifying being set on the agent's Config struct as Config.Serf*Config.MemberlistConfig.*.

The runtime tests were updated to test having these configurations in the hcl/json config and making sure they end up with the correct values in the RuntimeConfig. I then did a little manual testing to verify that the values get set properly in the agent's Config from the RuntimeConfig.

### CODE REVIEW

File: agent/config/default.go [hunk 1]
```diff
@@ -26,6 +26,10 @@ func DefaultRPCProtocol() (int, error) {
 // todo(fs): IMO, this should be the definitive default for all configurable values
 // todo(fs): and whatever is in here should clobber every default value. Hence, no sourcing.
 func DefaultSource() Source {
+	cfg := consul.DefaultConfig()
+	serfLAN := cfg.SerfLANConfig.MemberlistConfig
+	serfWAN := cfg.SerfWANConfig.MemberlistConfig
```
Comment (banks) [inline agent/config/default.go]: What's the significance of these moving from `DefaultConsulSource` to here? This seems like a better place but I don't recall offhand the details of how these two methods interact...

File: agent/config/runtime.go [hunk 2]
```diff
@@ -954,6 +946,160 @@ type RuntimeConfig struct {
...
+	// this value (more frequent) gossip messages are propagated across
+	// the cluster more quickly at the expense of increased bandwidth. This
+	// configuration only applies to LAN gossip communications
+	//
+	// The default is: 200ms
+	//
+	// hcl: gossip_lan { gossip_interval = duration}
+	GossipLANGossipInterval time.Duration
+
+	// GossipfLANGossipNodes is the number of random nodes to send gossip messages to
```
Comment (banks) [inline agent/config/runtime.go]: typo: extra `f` in the variable name.

File: agent/config/runtime.go [hunk 3]
```diff
@@ -954,6 +946,160 @@ type RuntimeConfig struct {
...
+	// per GossipInterval. Increasing this number causes the gossip messages to
+	// propagate across the cluster more quickly at the expense of increased
+	// bandwidth. This configuration only applies to LAN gossip communications
+	//
+	// The default is: 3
+	//
+	// hcl: gossip_lan { gossip_nodes = int}
```
Comment (banks) [inline agent/config/runtime.go]: nit: space after `int`

File: agent/config/runtime.go [hunk 4]
```diff
@@ -954,6 +946,160 @@ type RuntimeConfig struct {
...
+	// still alive.
+	//
+	// The default is: 4
+	//
+	// hcl: gossip_lan { suspicion_mult = int}
+	GossipLANSuspicionMult int
+
+	// GossipLANRetransmitMult is the multiplier for the number of retransmissions
+	// that are attempted for messages broadcasted over gossip. This
+	// configuration only applies to WAN gossip communications. The actual
```
Comment (banks) [inline agent/config/runtime.go]: LAN variable but you said "only applies to WAN"

File: agent/config/runtime.go [hunk 5]
```diff
@@ -954,6 +946,160 @@ type RuntimeConfig struct {
...
+	// this value (more frequent) gossip messages are propagated across
+	// the cluster more quickly at the expense of increased bandwidth. This
+	// configuration only applies to WAN gossip communications
+	//
+	// The default is: 200ms
+	//
+	// hcl: gossip_wan { gossip_interval = duration}
+	GossipWANGossipInterval time.Duration
+
+	// GossipfWANGossipNodes is the number of random nodes to send gossip messages to
```
Comment (banks) [inline agent/config/runtime.go]: Extra f again

File: agent/config/runtime_test.go [hunk 6]
```diff
@@ -286,14 +286,14 @@ func TestConfigFlagsAndEdgecases(t *testing.T) {
...
-				rt.ConsulSerfWANProbeTimeout = 100 * time.Millisecond
-				rt.ConsulSerfWANSuspicionMult = 3
+				rt.GossipLANGossipInterval = 100 * time.Millisecond
+				rt.GossipLANProbeInterval = 100 * time.Millisecond
+				rt.GossipLANProbeTimeout = 100 * time.Millisecond
+				rt.GossipLANSuspicionMult = 3
+				rt.GossipWANGossipInterval = 100 * time.Millisecond
+				rt.GossipWANProbeInterval = 100 * time.Millisecond
+				rt.GossipWANProbeTimeout = 100 * time.Millisecond
+				rt.GossipWANSuspicionMult = 3
```
Comment (banks) [inline agent/config/runtime_test.go]: Why don't we have tests for `Gossip*Nodes` and `Gossip*RetransmitMult` here again?

File: website/source/docs/agent/options.html.md [hunk 7]
```diff
@@ -883,6 +883,72 @@ Consul will not enable TLS for the HTTP API unless the `https` port has been ass
 * <a name="disable_keyring_file"></a><a href="#disable_keyring_file">`disable_keyring_file`</a> - Equivalent to the
   [`-disable-keyring-file` command-line flag](#_disable_keyring_file).
 
+* <a name="gossip_lan"></a><a href="#gossip_lan">`gossip_lan`</a> - (Advanced) This object contains a number of sub-keys
+  which can be set to tune the LAN gossip communications. These are only provided for users running especially large 
+  clusters that need fine tuning and are prepared to spend significant effort correctly tuning them for their
+  environment and workload.
```
Comment (banks) [inline website/source/docs/agent/options.html.md]: This is great, maybe add a sentence: "The default values are most appropriate in almost all deployments." to hammer it home.

File: website/source/docs/agent/options.html.md [hunk 8]
```diff
@@ -883,6 +883,72 @@ Consul will not enable TLS for the HTTP API unless the `https` port has been ass
...
+    zero, non-piggyback gossip is disabled. By lowering this value (more frequent) gossip messages are propagated
+    across the cluster more quickly at the expense of increased bandwidth.
+  
+  * <a name="probe_interval"></a><a href="#probe_interval">`probe_interval`</a> - The interval between random node
+    probes. Setting this lower (more frequent) will cause the cluster to detect failed nodes more quickly 
+    at the expense of increased bandwidth usage.
+  
+  * <a name="probe_timeout"></a><a href="#probe_timeout">`probe_timeout`</a> - The timeout to wait for an ack from
+    a probed node before assuming it is unhealthy. This should be set to 99-percentile of RTT (round-trip time) on
+     your network.
```
Comment (banks) [inline website/source/docs/agent/options.html.md]: Should we specify the default? Also maybe the language is confusing "This should be set to <variable that might be unique to your DC>" implies that users _should_ fiddle with this when we don't really want to encourage that.

Maybe instead something like "This should be at least the 99-percentile of RTT in you network. The default is 500ms and is a conservative value suitable for almost all realistic deployments."

File: website/source/docs/agent/options.html.md [hunk 9]
```diff
@@ -883,6 +883,72 @@ Consul will not enable TLS for the HTTP API unless the `https` port has been ass
...
+  
+  * <a name="retransmit_mult"></a><a href="#retransmit_mult">`retransmit_mult`</a> - The multiplier for the number 
+    of retransmissions that are attempted for messages broadcasted over gossip. The number of retransmits is scaled
+    using this multiplier and the cluster size. The higher the multiplier, the more likely a failed broadcast is to
+    converge at the expense of increased bandwidth.
```
Comment (banks) [inline website/source/docs/agent/options.html.md]: Default

**Comment by mkeeler** [comment]
Some of this might require a more in depth explanation of how the config is loaded.

Our configuration build (agent/config/builder.go) has a concept of user configuration as well as injected configurations. When given config fragments and config files. Our base configuration fragments to inject are located in agent/config/default.go.

How the builder sets things up is:

DefaultSource
DevSource (only if -dev mode is enabled)
User Defined Sources
NonUserSource
DefaultConsulSource
DefaultEnterpriseSource
DefaultVersionSource
DevConsulSource (if -dev mode is enabled)

Why we have NonUserSource and DefaultConsulSource I am not entirely sure. They both just override anything configured by the user previously to force specific values.

DevSource just overrides default configuration that is setup in DefaultSource and should only contain those items that it specifically wants to override.

Previously the gossip tuneables were in the DefaultConsulSource and thus not user modifiable even though they were technically valid configuration entries. They also existed in DevConsulSource to override them to better values for a -dev environment.

This PR shuffles things around a bit. No longer are the tuneables under the Consul configuration but we also want them to be user modifiable so they were put into the DefaultSource. This sets up sane defaults. A few of them that were previously overridden in DevConsulSource were moved to DevSource to get the override but to still allow user modification.

File: agent/config/default.go [hunk 10]
```diff
@@ -111,6 +132,18 @@ func DevSource() Source {
 		log_level = "DEBUG"
 		server = true
 
+		gossip_lan = {
+			gossip_interval = "100ms"
+			probe_interval = "100ms"
+			probe_timeout = "100ms"
+			suspicion_mult = 3
```
[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/config/default.go]: See my other comment. We don't want to do that here. Only items specifically overridden for dev mode go here.

File: agent/config/default.go (hunk 1 above)
Comment (mkeeler) [reply agent/config/default.go]: In depth explanation in my other comment.

DefaultConsulSource = non user modifiable
DefaultSource = user modifiable

File: agent/config/runtime_test.go (hunk 6 above)
Comment (mkeeler) [reply agent/config/runtime_test.go]: That test is just for -dev mode overrides. I guess we could through validation for some other values not being overridden but other tests already cover that.

File: website/source/docs/agent/options.html.md [hunk 11]
```diff
@@ -883,6 +883,74 @@ Consul will not enable TLS for the HTTP API unless the `https` port has been ass
 * <a name="disable_keyring_file"></a><a href="#disable_keyring_file">`disable_keyring_file`</a> - Equivalent to the
   [`-disable-keyring-file` command-line flag](#_disable_keyring_file).
 
+* <a name="gossip_lan"></a><a href="#gossip_lan">`gossip_lan`</a> - (Advanced) This object contains a number of sub-keys
```
Comment (pearkes) [inline website/source/docs/agent/options.html.md]: I'd add:

> Tuning these improperly can cause Consul to fail in unexpected ways.

Or something similarly strong language wise about potentially shooting yourself in the foot.

Comment (pearkes) [reply website/source/docs/agent/options.html.md]: And maybe even bold that line 😄

File: website/source/docs/agent/options.html.md (hunk 7 above)
Comment (pearkes) [reply website/source/docs/agent/options.html.md]: Ha I just repeated this even after that change. 😂

**Comment by mkeeler** [comment]
Docs update and bolded where appropriate.

**Pull request merged**